### PR TITLE
Constants reviver

### PR DESCRIPTION
**Pull request by matanlurey**
Partial support of https://github.com/dart-lang/source_gen/issues/178.

Still missing as of this PR:
* Any sort of sugar for `enum`, though it should work out of the box anyway.
* Error recovery: We just return `null` when something is not resolvable, ideally I'd like to either throw or return a specialized instance (`Unrevivable`) that explains why not (usually due to a private constructor, or not being able to trace how the object was created

Good starting point though I think.

### CODE REVIEW

File: lib/src/revive.dart [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/constant/value.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/src/dart/constant/value.dart';
+
+import 'utils.dart';
+
+/// Base interface for a meta-analyzed [DartObject].
+abstract class Revivable {}
```
Comment (kevmoo) [inline lib/src/revive.dart]: Weird to have an empty base class w/ no members – and one implementation?

File: test/constants_test.dart [hunk 2]
```diff
@@ -128,4 +128,52 @@ void main() {
       expect($deprecated.instanceOf(check), isTrue, reason: '$deprecated');
     });
   });
+
+  group('Reviable', () {
```
Comment (kevmoo) [inline test/constants_test.dart]: missing a `v`

File: lib/src/revive.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+FieldElement _findPublicConstField(DartObject object, ClassElement clazz) =>
+    clazz.fields.firstWhere(
+        (f) => f.isConst && f.isPublic && f.computeConstantValue() == object,
+        orElse: () => null);
+
+ConstructorInvocation _findPublicConstConstructor(
+  DartObject object,
+  ClassElement clazz,
+) {
+  final invocation = (object as DartObjectImpl).getInvocation();
```
Comment (kevmoo) [inline lib/src/revive.dart]: `getInvocation` may return null – need to handle that!

**Comment by matanlurey** [comment]
@kevmoo PTAL, lots of changes.

File: lib/src/revive.dart (hunk 3 above)
Comment (matanlurey) [inline lib/src/revive.dart]: Obsolete.

File: lib/src/revive.dart [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import 'package:analyzer/src/dart/constant/value.dart';
+
+import 'utils.dart';
+
+/// Base interface for a meta-analyzed [DartObject].
+abstract class Revivable {}
+
+/// Returns a revivable instance of [object].
+///
+/// Optionally specify the [clazz] type that contains the constructor.
```
Comment (matanlurey) [inline lib/src/revive.dart]: Obsolete.

File: lib/src/constants.dart [hunk 5]
```diff
@@ -151,6 +176,14 @@ class _Constant implements ConstantReader {
...
 
   @override
+  dynamic get anyValue =>
+      _object.toBoolValue() ??
+      _object.toIntValue() ??
+      _object.toStringValue() ??
+      _object.toListValue() ??
+      _object.toMapValue();
+
+  @override
```
Comment (kevmoo) [inline lib/src/constants.dart]: You're missing `symbol` here – it's supported on `DartObject`

You'll just want to wrap it in `new Symbol` if it returns a String.

Comment (kevmoo) [inline lib/src/constants.dart]: Also missing `double`

File: lib/src/revive.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+/// Returns a revivable instance of [object].
+///
+/// Optionally specify the [library] type that contains the reference.
+///
+/// Returns [null] if not revivable.
+Revivable reviveInstance(DartObject object, [LibraryElement library]) {
+  library ??= object.type.element.library;
+  var url = Uri.parse(urlOfElement(object.type.element));
+  final clazz = object?.type?.element as ClassElement;
+  for (final e in clazz.fields.where(
```
Comment (natebosch) [inline lib/src/revive.dart]: why does this need to be a for loop if the first line is a return? Is this a typical pattern to handle the 0 or first case?

would this be more readable as

```
var matchingFields = clazz.fields.where((f) => ...);
if (matchingFields.isNotEmpty) {
return new Revivable._(source: url, accessor: matchinfFields.first.name);
}

File: lib/src/revive.dart [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/src/dart/constant/value.dart';
+
+import 'utils.dart';
+
+/// Returns a revivable instance of [object].
+///
+/// Optionally specify the [library] type that contains the reference.
+///
+/// Returns [null] if not revivable.
```
Comment (natebosch) [inline lib/src/revive.dart]: Do we have known cases we know won't work? Could we expand this?

File: lib/src/revive.dart (hunk 6 above)
Comment (kevmoo) [inline lib/src/revive.dart]: Or just `firstWhere` – or even better – `singeWhere` – assuming more than one would be a bug

**Pull request merged**